### ISSUE TITLE
add a insertBeforeAndAfter

### ISSUE THREAD

**Issue by clankill3r** [jboss-javassist/javassist#292]
This method would be really welcome:
`void insertBeforeAndAfter(String srcBefore, String sourceAfter)`

Cause this is not possible for example"
```java
cm.insertBefore("{");
cm.insertAfter("}");
```
While it would have been valid code.

**Comment by chibash** [comment]
I cannot see any good algorithm to determine whether these arguments are valid or not.  Can you?

**Comment by clankill3r** [comment]
Sorry I don't understand the question.

**Comment by chibash** [comment]
Ah, I intended to say that I didn't know how to implement your idea.
Since the source code of the method body is not available,  making `insertBeforeAndAfter` work for arbitrary cases would be more difficult than you expect.

**Comment by chibash** [comment]
It might be good to add a bit more comment.
Javassist modifies the bytecode without its source code as if it *virtually* allows the users to modify the source code.  Unfortunately, this is not real.  So many things are difficult to enable although it looks very easy for the users who misconceive that Javassist modifies and compiles the source code to obtain the modified byte code.  Your suggestion is the case as long as I can see.

**Comment by chibash** [comment]
Please note that Javassist never decompiles the bytecode to obtain the source code.
The Java bytecode preserves many semantic features in the source code but obtaining the exact source code by decompilation is very difficult.  Lot's of corner cases.  Reproducing loop structures needs complex algorithms and I don't know the perfect one. etc.

**Comment by clankill3r** [comment]
I probably know too little, but my idea was to combine the `String srcBefore, String sourceAfter` into one string, see if that can be compiled into byte code. Then prepend the byte code of the srcBefore part and append the bytecode of the sourceAfter part.

**Comment by chibash** [comment]
Oh, "Then prepend the byte code of the srcBefore part and append the bytecode of the sourceAfter part." is not easy.  How can you split the compiled binary into the before part and the after part?  Please remember how the source code is compiled into bytecode.
I must say it is *not* impossible but needs lots of engineering.

**Comment by clankill3r** [comment]
I would but some source in between that has recognisable bytecode, that you later can search for and remove.

**Comment by chibash** [comment]
Yes, but the problem is the two bytecode sequences split at that recognizable bytecode would not work unless the local variable numbers (i.e. register numbers) are renumbered, the exception table is updated, the stack map table are reconstructed, etc.  These modifications will be possible but would need lots of engineering efforts.  We also have to come up with an algorithm to detect  whether valid source code is given to `insertBeforeAndAfter` and report an error when the code is invalid.

**Comment by blackr1234** [comment]
As the developer of the library @chibash said, it's technically challenging to accomplish what the OP wants. However, there's a workaround.

Consider the following code. It may not always work and you may end up checking a lot of things, but if you are not working on a very complex class it should just work.

```java
package code;

import javassist.ClassPool;
import javassist.CtClass;
import javassist.CtMethod;

public class Main3 {
    public static void main(String[] args) throws Exception {

        ClassPool cp = ClassPool.getDefault();
        CtClass cc = cp.get("code.Hello");
        insertBeforeAndAfter(
            cc.getDeclaredMethod("sayHello"),
            "{ " + "System.out.println(\"Calling the original method.\");",
            "System.out.println(\"Done calling the original method.\");" + "}");
        Class<?> c = cc.toClass();

        ((Hello) c.newInstance()).sayHello();
    }

    private static void insertBeforeAndAfter(CtMethod cm, String before, String after) throws Exception {

        final CtClass cc = cm.getDeclaringClass();
        final String originalMethodName = cm.getName();
        final String newMethodName = "$" + cm.getName();

        cm.setName(newMethodName);

        cc.addMethod(CtMethod.make(
            "public void " + originalMethodName + "() {"
          +     before
          +     newMethodName + "();"
          +     after
          + "}",
          cc));
    }
}

class Hello {
    public void sayHello() {
        System.out.println("hello~");
    }
}
```

**Comment by clankill3r** [comment]
@blackr1234 Thank you. I was about to work on this in the upcoming days. Your post comes at a welcome time 👍

**Comment by blackr1234** [comment]
> @blackr1234 Thank you. I was about to work on this in the upcoming days. Your post comes at a welcome time 👍

You're welcome.

I just updated my answer. It turns out that I can just clone a `CtMethod` and modify some of its parts, such as its method body.
By the way, I also find that there's such a method: `CtMethod#setWrappedBody(mbody, constParam)`. Its JavaDoc states that it can:

> Replace a method body with a new method body wrapping the given method.

I haven't given it a try but there's a chance it can achieve our goal too.
